### PR TITLE
[P2P] refactor: nonce deduper

### PR DESCRIPTION
**Pull request by bryanchriswhite**
### Summary generated by Reviewpad on 01 Jun 23 07:03 UTC
This pull request contains multiple diffs across different files. Changes include moving the nonce field from RainTreeMessage to PocketEnvelope protobuf types, formalizing the NonceDeduper type and moving it to the utils package, modifying function parameters and removing unnecessary code, adding a dependency on mempool, emitting telemetry events on duplicate messages, adding a capacity check for MaxNonces, removing MaxNonces from struct initializations and from config validation, renaming a test file to a different package, and updating the changelog with version 0.0.0.58 changes.

### Before

```mermaid
classDiagram
    class RainTreeMessage {
        <<protobuf>>
        +Level uint32
        +Data []byte
        +Nonce uint64
    }
    
    class PocketEnvelope {
        <<protobuf>>
        +Content *anypb.Any   
    }

    RainTreeMessage --* PocketEnvelope : serialized as `Data`
    
      
    class P2PModule {

    }

    class RainTreeRouter {
        
    }

    %% RainTreeRouter --> P2PModule
    %% P2PModule --> RainTreeRouter
    RainTreeRouter --o RainTreeMessage
    RainTreeRouter --* NonceDeduper
    P2PModule --o PocketEnvelope
```

### After

```mermaid
classDiagram
    class RainTreeMessage {
        <<protobuf>>
        +Level uint32
        +Data []byte
    }
    
    class PocketEnvelope {
        <<protobuf>>
        +Content *anypb.Any
        +Nonce uint64
    }

    RainTreeMessage --* PocketEnvelope : serialized as `Data`
    
    class P2PModule {

    }

    class RainTreeRouter {

    }

    %% RainTreeRouter --> P2PModule
    RainTreeRouter --o RainTreeMessage

    P2PModule --* NonceDeduper
    P2PModule --o PocketEnvelope
```

## Issue

Deliverables 1 & 3 of #762 

## Type of change

Please mark the relevant option(s):

- [ ] New feature, functionality or library
- [ ] Bug fix
- [x] Code health or cleanup
- [ ] Major breaking change
- [ ] Documentation
- [ ] Other 

## List of changes

- Moved `nonce` field from `RainTreeMessage` to `PocketEnvelope` protobuf types
- Formalize `NonceDeduper` (and related) types
- Move `NonceDeduper` (and related types) from p2p/raintree to p2p/utils pkg
- Move `NonceDeduper` usage out of `RainTreeRouter` and into `p2pModule`

## Testing

- [ ] `make develop_test`; if any code changes were made
- [ ] `make test_e2e` on [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any code changes were made
- [ ] `e2e-devnet-test` passes tests on [DevNet](https://pocketnetwork.notion.site/How-to-DevNet-ff1598f27efe44c09f34e2aa0051f0dd); if any code was changed
- [x] [Docker Compose LocalNet](https://github.com/pokt-network/pocket/blob/main/docs/development/README.md); if any major functionality was changed or introduced
- [x] [k8s LocalNet](https://github.com/pokt-network/pocket/blob/main/build/localnet/README.md); if any infrastructure or configuration changes were made

## Required Checklist

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added, or updated, [`godoc` format comments](https://go.dev/blog/godoc) on touched members (see: [tip.golang.org/doc/comment](https://tip.golang.org/doc/comment))
- [ ] I have tested my changes using the available tooling
- [ ] I have updated the corresponding CHANGELOG

### If Applicable Checklist

- [ ] I have updated the corresponding README(s); local and/or global
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added, or updated, [mermaid.js](https://mermaid-js.github.io) diagrams in the corresponding README(s)
- [ ] I have added, or updated, documentation and [mermaid.js](https://mermaid-js.github.io) diagrams in `shared/docs/*` if I updated `shared/*`README(s)

### CODE REVIEW

File: p2p/module.go
Comment (Olshansk) [inline p2p/module.go:355]:
```suggestion
		return fmt.Errorf("handling app data: %w", err)
```

File: p2p/module.go
Comment (Olshansk) [inline p2p/module.go:380]: Nice

File: p2p/module.go
Comment (Olshansk) [inline p2p/module.go:371]:
```suggestion
// checkNonce ensures the nonce has not been observed within the deuper's capacity
```

File: p2p/module.go
Comment (Olshansk) [inline p2p/module.go:374]:
```suggestion
func (m *p2pModule) checkNonce(nonce utils.Nonce) (bool, error) {
```

Comment (Olshansk) [inline p2p/module.go:374]: Can you rename `checkNonse` to something more "functional" like `wasNonceObserved` or `isDuplicateNonce` so there is no ambiguity as to what the boolean return value contains.

File: shared/messaging/proto/pocket_envelope.proto
Comment (Olshansk) [inline shared/messaging/proto/pocket_envelope.proto:11]:
```suggestion
  uint64 nonce = 2; // TECHDEBT: should this be the same as the nonce in `Transaction`?
```

File: p2p/module.go
Comment (Olshansk) [inline p2p/module.go:361]: I do not remember the exact reasoning. Your example/reasoning seems accurate, and I believe our attempt at making sure the nonce wasn't propagated was once of the reasons.

I believe one of the reasons was to pass a structure by value, rather than by reference, in case downstream functionality may try to modify it.

**Review by Olshansk — CHANGES_REQUESTED** [review]
Could you add your mermaid diagram in the PR description to the README?

**Comment by bryanchriswhite** [comment]
> Could you add your mermaid diagram in the PR description to the README?

See: https://github.com/pokt-network/pocket/pull/763#issuecomment-1558730255

File: p2p/module.go
Comment (bryanchriswhite) [inline p2p/module.go:393]:
```suggestion
		return true
	}
	return false
```

File: p2p/module.go
Comment (Olshansk) [inline p2p/module.go:354]: Consider logging this, at least in the debug level.

File: p2p/module.go
Comment (Olshansk) [inline p2p/module.go:362]:
```suggestion
	// NB: Explicitly constructing a new `PocketEnvelope` literal with content
```

File: p2p/module.go
Comment (Olshansk) [inline p2p/module.go:374]: Can we just do `return m.nonceDeduper.Push(nonce)`

File: p2p/module.go
Comment (Olshansk) [inline p2p/module.go:383]: Can you link to this PR w/ a comment in the discussion after we merge it in?

File: p2p/module.go
Comment (Olshansk) [inline p2p/module.go:385]: NIT: If you do `!contains` and return `false`, you wouldn't have to indent all the business logic.

**Review by Olshansk — CHANGES_REQUESTED** [review]

File: p2p/module.go
Comment (bryanchriswhite) [reply p2p/module.go:385]: Thanks for calling this out! :raised_hands: I hadn't taken the time to properly refactor this function. :sweat_smile:

File: p2p/module.go
Comment (bryanchriswhite) [reply p2p/module.go:374]: :thinking: We could. I would prefer just simplifying this definition:

```
// observeNonce adds the nonce to the deduper if it has not been observed.
func (m *p2pModule) observeNonce(nonce utils.Nonce) error {
	// Add the nonce to the deduper
	return m.nonceDeduper.Push(nonce)
}
```

I think having `#observeNonce()` and `#isNonceObserved()` encapsulate the details of interacting with the nonce deduper follows the [single-responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle). Even though the API of the nonce deduper must be known to the p2p module, it's not necessary to expose/interact with it more than necessary; giving every other piece of code which references the deduper an additional "reason to change".

Comment (bryanchriswhite) [reply p2p/module.go:374]: Perhaps a even better alternative would be to change `NonceDeduper` to no longer be an alias and them move these two methods to it instead. :thinking:

**Review by dylanlott — COMMENTED** [review]
Looks good to me, but I'm gonna hold off on approving until @Olshansk. Pushing nonce deduplication down to the p2p module's level is absolutely the right call. 👍

File: p2p/module.go
Comment (Olshansk) [reply p2p/module.go:374]: Completely on the same page regarding the wrapper :)

**Review by Olshansk — APPROVED** [review]
Thanks for bearing through all the reviews. Lgtm!

File: shared/messaging/proto/pocket_envelope.proto
Comment (bryanchriswhite) [inline shared/messaging/proto/pocket_envelope.proto:11]: :+1: I'm not super familiar with the background here, this is a comment that has been transported and transformed across history. AFAIK discussion #278 has the most detail on this:

```suggestion
  uint64 nonce = 2; // DISCUSS(#278): should this be the same as the nonce in `Transaction`?
```

File: p2p/CHANGELOG.md
Comment (bryanchriswhite) [inline p2p/CHANGELOG.md:10]:
```suggestion
## [0.0.0.53] - 2023-06-01
```

File: shared/CHANGELOG.md
Comment (bryanchriswhite) [inline shared/CHANGELOG.md:10]:
```suggestion
## [0.0.0.58] - 2023-06-01
```

**Pull request merged**
